sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/job_properties.hpp

```diff
@@ -13,12 +22,116 @@ namespace Azure { namespace Storage {
     DirectoryDownload = 3,
   };
 
+  enum class JobStatus
+  {
+    InProgress,
+    Paused,
+    Successful,
+    Failed,
+    PartiallySuccessful,
+    Cancelled,
+  };
+
   struct JobProperties final
   {
-    std::string JobId;
+    std::string Id;
     std::string SourceUrl;
     std::string DestinationUrl;
-    TransferType Type;
+    TransferType Type = static_cast<TransferType>(std::underlying_type_t<TransferType>(0));
+    std::shared_future<JobStatus> WaitHandle;
```

Comment:
I'm fine with std::shared_future for waiting on completion in C++, but it's not support in C or P/Invoke. 

With JobStatus as parameter, if the job is failed, customer won't be able to get the fail message from here.  It's just feeling strange to get completion and detailed error message in two different places.

Reply:
> I'm fine with std::shared_future for waiting on completion in C++, but it's not support in C or P/Invoke.

I'd like to focus on C++ ATM. We couldn't use the same API interface anyway in we were going to work on C, like `std::string` isn't available there either.

> With JobStatus as parameter, if the job is failed, customer won't be able to get the fail message from here. It's just feeling strange to get completion and detailed error message in two different places.

There could be multiple error messages for one job. In the worst case, errors cannot fit into memory. We need to be able to handle this case. I'm guessing callback is the best choice.

Do you have other suggestions here?

Reply:
We can leave it here for now. It's an interface design, we'd need a  principle for this.  I agree that the callback is a good choice to report all detailed errors, may be good to have a summary error here as the transfer option is optional and for some simple transfer customer may not bother to specify a callback. Anyway, we can keep it here for now.